<a href="https://colab.research.google.com/github/veernaik/Data_Science_16014324004_TY/blob/main/Data_Science_exp3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [8]:
import pandas as pd
import math

df = pd.read_csv("weatherAUS.csv")

attributes = [
    "MinTemp",
    "MaxTemp",
    "Rainfall",
    "Humidity9am",
    "Pressure9am"
]

results = []

for column in attributes:

    data = []

    for value in df[column]:
        if pd.notna(value):
            data.append(value)

    n = len(data)

    total = 0

    for value in data:
        total = total + value

    mean_before = total / n

    sorted_data = sorted(data)

    if n % 2 == 0:
        median_before = (
            sorted_data[n//2 - 1] +
            sorted_data[n//2]
        ) / 2
    else:
        median_before = sorted_data[n//2]

    frequencies = {}

    for value in data:
        if value in frequencies:
            frequencies[value] = frequencies[value] + 1
        else:
            frequencies[value] = 1

    highest_frequency = 0
    mode_before = None

    for value in frequencies:
        if frequencies[value] > highest_frequency:
            highest_frequency = frequencies[value]
            mode_before = value

    minimum = sorted_data[0]
    maximum = sorted_data[-1]

    data_range = maximum - minimum

    k = math.ceil(
        1 + 3.322 * math.log10(n)
    )

    class_width = data_range / k

    bins = []

    for i in range(k + 1):
        bins.append(
            minimum + i * class_width
        )

    frequencies_list = []
    midpoints = []

    for i in range(k):

        lower = bins[i]
        upper = bins[i + 1]

        frequency = 0

        for value in data:

            if i == k - 1:
                if lower <= value <= upper:
                    frequency = frequency + 1
            else:
                if lower <= value < upper:
                    frequency = frequency + 1

        frequencies_list.append(frequency)

        midpoint = (lower + upper) / 2
        midpoints.append(midpoint)

    total_frequency = 0

    for frequency in frequencies_list:
        total_frequency = total_frequency + frequency

    grouped_mean = 0

    for i in range(len(midpoints)):
        grouped_mean = grouped_mean + (
            midpoints[i] *
            frequencies_list[i]
        )

    grouped_mean = grouped_mean / total_frequency

    median_position = total_frequency / 2

    cumulative_frequency = 0
    median_class = 0

    for i in range(len(frequencies_list)):

        cumulative_frequency = (
            cumulative_frequency +
            frequencies_list[i]
        )

        if cumulative_frequency >= median_position:
            median_class = i
            break

    previous_frequency = 0

    for i in range(median_class):
        previous_frequency = (
            previous_frequency +
            frequencies_list[i]
        )

    lower_boundary = bins[median_class]

    median_class_frequency = (
        frequencies_list[median_class]
    )

    grouped_median = lower_boundary + (
        (
            (
                median_position -
                previous_frequency
            )
            / median_class_frequency
        )
        * class_width
    )

    modal_class = 0

    for i in range(len(frequencies_list)):

        if frequencies_list[i] > frequencies_list[modal_class]:
            modal_class = i

    lower_modal = bins[modal_class]

    if modal_class == 0:
        previous_modal_frequency = 0
    else:
        previous_modal_frequency = (
            frequencies_list[modal_class - 1]
        )

    if modal_class == len(frequencies_list) - 1:
        next_modal_frequency = 0
    else:
        next_modal_frequency = (
            frequencies_list[modal_class + 1]
        )

    modal_frequency = frequencies_list[modal_class]

    denominator = (
        (2 * modal_frequency)
        - previous_modal_frequency
        - next_modal_frequency
    )

    if denominator == 0:
        grouped_mode = lower_modal
    else:
        grouped_mode = lower_modal + (
            (
                (
                    modal_frequency -
                    previous_modal_frequency
                )
                / denominator
            )
            * class_width
        )

    results.append([
        column,
        n,
        mean_before,
        median_before,
        mode_before,
        minimum,
        maximum,
        data_range,
        k,
        class_width,
        grouped_mean,
        grouped_median,
        grouped_mode
    ])

results_df = pd.DataFrame(
    results,
    columns=[
        "Attribute",
        "Number of Values",
        "Mean Before Sturges",
        "Median Before Sturges",
        "Mode Before Sturges",
        "Minimum",
        "Maximum",
        "Range",
        "Sturges Classes",
        "Class Width",
        "Mean After Sturges",
        "Median After Sturges",
        "Mode After Sturges"
    ]
)

print(results_df)

     Attribute  Number of Values  Mean Before Sturges  Median Before Sturges  \
0      MinTemp            143975            12.194034                   12.0   
1      MaxTemp            144199            23.221348                   22.6   
2     Rainfall            142199             2.360918                    0.0   
3  Humidity9am            142806            68.880831                   70.0   
4  Pressure9am            130395          1017.649940                 1017.6   

   Mode Before Sturges  Minimum  Maximum  Range  Sturges Classes  Class Width  \
0                 11.0     -8.5     33.9   42.4               19     2.231579   
1                 20.0     -4.8     48.1   52.9               19     2.784211   
2                  0.0      0.0    371.0  371.0               19    19.526316   
3                 99.0      0.0    100.0  100.0               19     5.263158   
4               1016.4    980.5   1041.0   60.5               18     3.361111   

   Mean After Sturges  Median Af

Inference:- From the results, MinTemp and MaxTemp have mean values of 12.19°C and 23.22°C, while Rainfall has a mean of 2.36 mm and mode of 0 mm.
The Sturges’ rule gives around 18–19 classes, and the grouped values are close to the original values.

In [10]:
import pandas as pd
import numpy as np

df = pd.read_csv("weatherAUS.csv")

attributes = [
    "MinTemp",
    "MaxTemp",
    "Rainfall",
    "Humidity9am",
    "Pressure9am"
]

results = []

for column in attributes:

    data = df[column].dropna()

    frequency = len(data)

    if column == "MinTemp":

        values = []

        for value in data:
            values.append(value)

        n = len(values)

        total = 0

        for value in values:
            total = total + value

        mean = total / n

        squared_difference_sum = 0

        for value in values:
            difference = value - mean
            squared_difference_sum = (
                squared_difference_sum +
                difference * difference
            )

        variance = squared_difference_sum / n

        standard_deviation = np.sqrt(variance)

        sorted_data = sorted(values)

        n = len(sorted_data)

        if n % 2 == 0:
            lower_half = sorted_data[:n//2]
            upper_half = sorted_data[n//2:]
        else:
            lower_half = sorted_data[:n//2]
            upper_half = sorted_data[n//2 + 1:]

        lower_n = len(lower_half)

        if lower_n % 2 == 0:
            q1 = (
                lower_half[lower_n//2 - 1] +
                lower_half[lower_n//2]
            ) / 2
        else:
            q1 = lower_half[lower_n//2]

        upper_n = len(upper_half)

        if upper_n % 2 == 0:
            q3 = (
                upper_half[upper_n//2 - 1] +
                upper_half[upper_n//2]
            ) / 2
        else:
            q3 = upper_half[upper_n//2]

        iqr = q3 - q1

    else:

        variance = data.var()
        standard_deviation = data.std()
        iqr = data.quantile(0.75) - data.quantile(0.25)

    results.append([
        column,
        frequency,
        variance,
        standard_deviation,
        iqr
    ])

results_df = pd.DataFrame(
    results,
    columns=[
        "Attribute",
        "Frequency",
        "Variance",
        "Standard Deviation",
        "Interquartile Range"
    ]
)

print(results_df)

     Attribute  Frequency    Variance  Standard Deviation  Interquartile Range
0      MinTemp     143975   40.940454            6.398473                  9.3
1      MaxTemp     144199   50.680856            7.119049                 10.3
2     Rainfall     142199   71.877497            8.478060                  0.8
3  Humidity9am     142806  362.109100           19.029164                 26.0
4  Pressure9am     130395   50.502773            7.106530                  9.5


Inference:- The frequency of the attributes is high, ranging from 130395 to 144199, showing that most of the data is available for analysis.
The variation is different for each attribute, with Rainfall showing higher spread, while the IQR represents the spread of the middle 50% of the data.

In [11]:
data = [10, 12, 15, 12, 18, 20, 12, 25, 30, 15]

n = len(data)

total = 0

for value in data:
    total = total + value

mean = total / n

sorted_data = sorted(data)

if n % 2 == 0:
    median = (
        sorted_data[n//2 - 1] +
        sorted_data[n//2]
    ) / 2
else:
    median = sorted_data[n//2]

frequencies = {}

for value in data:
    if value in frequencies:
        frequencies[value] = frequencies[value] + 1
    else:
        frequencies[value] = 1

highest_frequency = 0
mode = None

for value in frequencies:
    if frequencies[value] > highest_frequency:
        highest_frequency = frequencies[value]
        mode = value

variance_sum = 0

for value in data:
    difference = value - mean
    variance_sum = variance_sum + difference * difference

variance = variance_sum / n

standard_deviation = math.sqrt(variance)

if n % 2 == 0:
    lower_half = sorted_data[:n//2]
    upper_half = sorted_data[n//2:]
else:
    lower_half = sorted_data[:n//2]
    upper_half = sorted_data[n//2 + 1:]

lower_n = len(lower_half)

if lower_n % 2 == 0:
    q1 = (
        lower_half[lower_n//2 - 1] +
        lower_half[lower_n//2]
    ) / 2
else:
    q1 = lower_half[lower_n//2]

upper_n = len(upper_half)

if upper_n % 2 == 0:
    q3 = (
        upper_half[upper_n//2 - 1] +
        upper_half[upper_n//2]
    ) / 2
else:
    q3 = upper_half[upper_n//2]

iqr = q3 - q1

print("Frequency:", n)
print("Mean:", mean)
print("Median:", median)
print("Mode:", mode)
print("Variance:", variance)
print("Standard Deviation:", standard_deviation)
print("Q1:", q1)
print("Q3:", q3)
print("Interquartile Range:", iqr)

Frequency: 10
Mean: 16.9
Median: 15.0
Mode: 12
Variance: 37.49000000000001
Standard Deviation: 6.122907805936654
Q1: 12
Q3: 20
Interquartile Range: 8


Inference:- For the given sample data, the mean is 16.9, median is 15 and mode is 12, showing that the data values are mainly concentrated around this range.
The variance is 37.49, standard deviation is 6.12 and IQR is 8, indicating the amount of variation present in the sample.

In [12]:
import pandas as pd
import math

df = pd.read_csv("weatherAUS.csv")

attributes = [
    "MinTemp",
    "MaxTemp",
    "Rainfall",
    "Humidity9am",
    "Pressure9am"
]

results = []

for column in attributes:

    data = []

    for value in df[column]:
        if pd.notna(value):
            data.append(value)

    n = len(data)

    sorted_data = sorted(data)

    minimum = sorted_data[0]
    maximum = sorted_data[-1]

    data_range = maximum - minimum

    k = math.ceil(
        1 + 3.322 * math.log10(n)
    )

    class_width = data_range / k

    bins = []

    for i in range(k + 1):
        bins.append(
            minimum + i * class_width
        )

    frequencies = []
    midpoints = []

    for i in range(k):

        lower = bins[i]
        upper = bins[i + 1]

        frequency = 0

        for value in data:

            if i == k - 1:

                if lower <= value <= upper:
                    frequency = frequency + 1

            else:

                if lower <= value < upper:
                    frequency = frequency + 1

        frequencies.append(frequency)

        midpoint = (lower + upper) / 2

        midpoints.append(midpoint)

    total_frequency = 0

    for frequency in frequencies:
        total_frequency = total_frequency + frequency

    grouped_mean_sum = 0

    for i in range(k):

        grouped_mean_sum = (
            grouped_mean_sum +
            midpoints[i] * frequencies[i]
        )

    grouped_mean = (
        grouped_mean_sum /
        total_frequency
    )

    variance_sum = 0

    for i in range(k):

        difference = (
            midpoints[i] -
            grouped_mean
        )

        variance_sum = (
            variance_sum +
            frequencies[i] *
            difference *
            difference
        )

    grouped_variance = (
        variance_sum /
        total_frequency
    )

    grouped_standard_deviation = math.sqrt(
        grouped_variance
    )

    median_position = total_frequency / 2

    cumulative_frequency = 0
    median_class = 0

    for i in range(k):

        cumulative_frequency = (
            cumulative_frequency +
            frequencies[i]
        )

        if cumulative_frequency >= median_position:

            median_class = i
            break

    previous_frequency = 0

    for i in range(median_class):

        previous_frequency = (
            previous_frequency +
            frequencies[i]
        )

    median_lower = bins[median_class]

    median_frequency = frequencies[median_class]

    grouped_median = (
        median_lower +
        (
            (
                median_position -
                previous_frequency
            )
            / median_frequency
        )
        * class_width
    )

    modal_class = 0

    for i in range(k):

        if frequencies[i] > frequencies[modal_class]:
            modal_class = i

    modal_lower = bins[modal_class]

    modal_frequency = frequencies[modal_class]

    if modal_class == 0:
        previous_modal_frequency = 0
    else:
        previous_modal_frequency = frequencies[
            modal_class - 1
        ]

    if modal_class == k - 1:
        next_modal_frequency = 0
    else:
        next_modal_frequency = frequencies[
            modal_class + 1
        ]

    denominator = (
        (2 * modal_frequency)
        - previous_modal_frequency
        - next_modal_frequency
    )

    if denominator == 0:

        grouped_mode = modal_lower

    else:

        grouped_mode = (
            modal_lower +
            (
                (
                    modal_frequency -
                    previous_modal_frequency
                )
                / denominator
            )
            * class_width
        )

    q1_position = total_frequency / 4
    q3_position = 3 * total_frequency / 4

    cumulative_frequency = 0
    q1_class = 0

    for i in range(k):

        cumulative_frequency = (
            cumulative_frequency +
            frequencies[i]
        )

        if cumulative_frequency >= q1_position:

            q1_class = i
            break

    previous_q1_frequency = 0

    for i in range(q1_class):

        previous_q1_frequency = (
            previous_q1_frequency +
            frequencies[i]
        )

    q1 = (
        bins[q1_class] +
        (
            (
                q1_position -
                previous_q1_frequency
            )
            / frequencies[q1_class]
        )
        * class_width
    )

    cumulative_frequency = 0
    q3_class = 0

    for i in range(k):

        cumulative_frequency = (
            cumulative_frequency +
            frequencies[i]
        )

        if cumulative_frequency >= q3_position:

            q3_class = i
            break

    previous_q3_frequency = 0

    for i in range(q3_class):

        previous_q3_frequency = (
            previous_q3_frequency +
            frequencies[i]
        )

    q3 = (
        bins[q3_class] +
        (
            (
                q3_position -
                previous_q3_frequency
            )
            / frequencies[q3_class]
        )
        * class_width
    )

    grouped_iqr = q3 - q1

    results.append([
        column,
        total_frequency,
        k,
        class_width,
        grouped_mean,
        grouped_median,
        grouped_mode,
        grouped_variance,
        grouped_standard_deviation,
        q1,
        q3,
        grouped_iqr
    ])

results_df = pd.DataFrame(
    results,
    columns=[
        "Attribute",
        "Frequency",
        "Sturges Classes",
        "Class Width",
        "Grouped Mean",
        "Grouped Median",
        "Grouped Mode",
        "Grouped Variance",
        "Grouped Standard Deviation",
        "Q1",
        "Q3",
        "Grouped IQR"
    ]
)

print(results_df)

     Attribute  Frequency  Sturges Classes  Class Width  Grouped Mean  \
0      MinTemp     143975               19     2.231579     12.192228   
1      MaxTemp     144199               19     2.784211     23.227782   
2     Rainfall     142199               19    19.526316     10.693617   
3  Humidity9am     142806               19     5.263158     68.829188   
4  Pressure9am     130395               18     3.361111   1017.650608   

   Grouped Median  Grouped Mode  Grouped Variance  Grouped Standard Deviation  \
0       12.032902     11.619618         41.307782                    6.427113   
1       22.648564     19.816402         51.360489                    7.166623   
2       10.060154      9.866866         48.035646                    6.930775   
3       69.909734     62.588956        359.941416                   18.972122   
4     1017.660439   1017.895295         51.532499                    7.178614   

            Q1           Q3  Grouped IQR  
0     7.567931    16.885038    

Inference:- The grouped data gives mean values close to the original data, such as 12.19 for MinTemp, 23.23 for MaxTemp and 68.83 for Humidity9am.
The grouped statistical measures show the distribution and spread of the data after dividing it into different classes.